evm/src/assets/sharedComponents/utils.sol

```diff
@@ -0,0 +1,27 @@
+// SPDX-License-Identifier: Apache 2
+
+pragma solidity ^0.8.25;
+
+import {SafeERC20} from "@openzeppelin/token/ERC20/utils/SafeERC20.sol";
+import {IERC20} from "@openzeppelin/token/ERC20/IERC20.sol";
+
+/**
+ * Payment to the target failed.
+ */
+error PaymentFailure(address target);
+
+using SafeERC20 for IERC20;
+
+function transferTokens(address token, address to, uint256 amount) {
+  if (token == address(0))
+    _transferEth(to, amount);
+  else
+    IERC20(token).safeTransfer(to, amount);
+}
+
+function _transferEth(address to, uint256 amount) {
+  if (amount == 0) return;
+
+  (bool success, ) = to.call{value: amount}(new bytes(0));
```

Comment:
This opens us up to gas abuse, though I think the right place to worry about this is in the SDK when composing the transaction.

Reply:
One thing I think we could do but I don't know if it's worth it is have the gas limit be an override provided by the tx sender.